[Docs] Move Message definition to separate section, add extra options (#6947)
PyCQA/pylint: doc/development_guide/how_tos/custom_checkers.rst

    environment variable or by adding the ``my_plugin.py``
    file to the ``pylint/checkers`` directory if running from source.

Defining a Message
------------------

Pylint message is defined using the following format::

   msgs = {
       "message-id": (
           "displayed-message", "message-symbol", "message-help"
       )
   }

* The ``message-id`` should be a 4-digit number,
  prefixed with a **message category**.
  There are multiple message categories,
  these being ``C``, ``W``, ``E``, ``F``, ``R``,
  standing for ``Convention``, ``Warning``, ``Error``, ``Fatal`` and ``Refactoring``.
  The 4 digits should not conflict with existing checkers

>   and the first 2 digits should consistent across the checker (except shared messages).

Author: This is just copy-past for current documentation.

>   and the first 2 digits should consistent across the checker (except shared messages).

Reviewer: I did a lot of work on the documentation recently (not on the dev documentation yet though as you see 😄) and there sure are part that can be reorganized or removed altogether.  Not being too verbose and staying focused is a good thing Don't feel obligated to clean up, but we can remove old documentation that is not required if necessary.

>   The 4 digits should not conflict with existing checkers
>   and the first 2 digits should consistent across the checker (except shared messages).

Reviewer: ```suggestion
```

>   and the first 2 digits should consistent across the checker (except shared messages).

Reviewer 2: Let's not change this before discussion on https://github.com/PyCQA/pylint/issues/5275 has resolved.

> * The ``displayed-message`` is used for displaying the message to the user,
>   once it is emitted.
> 
> * The ``message-symbol`` is an alias of the message id
>   and it can be used wherever the message id can be used.
> 
> * The ``message-help`` is used when calling ``pylint --help-msg``.
> 
> Optionally message can contain optional extra options::
> 
>    msgs = {
>        "message-id": (
>            "displayed-message", "message-symbol", "message-help",
>            {"extra_option_name": "extra_option_value"}
>        )
>    }

Reviewer: I think a full example would work well here, something understandable by everyone like:

```suggestion
* The ``displayed-message`` is used for displaying the message to the user,
  once it is emitted.

* The ``message-symbol`` is an alias of the message id
  and it can be used wherever the message id can be used.

* The ``message-help`` is used when calling ``pylint --help-msg``.

Optionally message can contain optional extra options::

   msgs = {
       "E0401": (
        "Unable to import %s",
        "import-error",
        "Used when pylint has been unable to import a module.",
        {"old_names": [("F0401", "old-import-error")]},
    ),

The message is then formatted using the ``args`` parameter from ``add_message`` i.e. in
``self.add_message("import-error", args=module_we_cant_import, node=importnode)``
    
```

> * The ``old_names`` option provides the way of specifying legacy ``message-id`` and
>   ``message-symbol`` of the message used in the past. The option is specified as a list

Reviewer: ```suggestion
* The ``old_names`` option permits to change the message id or symbol of a message without breaking the message control used on the old messages by users. The option is specified as a list
```

    environment variable or by adding the ``my_plugin.py``
    file to the ``pylint/checkers`` directory if running from source.

Defining a Message
------------------

Pylint message is defined using the following format::

>    msgs = {
>        "message-id": (
>            "displayed-message", "message-symbol", "message-help"
>        )
>    }

Reviewer: ```suggestion
   msgs = {
       "E0401": ( # message id
        "Unable to import %s", # template of displayed message
        "import-error", # message symbol
        "Used when pylint has been unable to import a module.",  # Message description
        { # Additional parameters:
             # message control support for the old names of the messages:
            "old_names": [("F0401", "old-import-error")] 
            "minversion": (3, 5), # No check under this version
            "maxversion": (3, 7), # No check above this version
        },
    ),

The message is then formatted using the ``args`` parameter from ``add_message`` i.e. in
``self.add_message("import-error", args=module_we_cant_import, node=importnode)``, the value in ``module_we_cant_import`` say "patglib`` will be interpolled and the final result will be:
``Unable to import patglib``
```

* The ``message-id`` should be a 4-digit number,
  prefixed with a **message category**.
  There are multiple message categories,
  these being ``C``, ``W``, ``E``, ``F``, ``R``,
  standing for ``Convention``, ``Warning``, ``Error``, ``Fatal`` and ``Refactoring``.
  The 4 digits should not conflict with existing checkers
  and the first 2 digits should consistent across the checker (except shared messages).

* The ``displayed-message`` is used for displaying the message to the user,
  once it is emitted.

* The ``message-symbol`` is an alias of the message id
  and it can be used wherever the message id can be used.

* The ``message-help`` is used when calling ``pylint --help-msg``.

Optionally message can contain optional extra options::

>    msgs = {
>        "E0401": (
>         "Unable to import %s",
>         "import-error",
>         "Used when pylint has been unable to import a module.",
>         {"old_names": [("F0401", "old-import-error")]},
>     ),
> 
> The message is then formatted using the ``args`` parameter from ``add_message`` i.e. in
> ``self.add_message("import-error", args=module_we_cant_import, node=importnode)``

Reviewer: ```suggestion
```
